Includes changes for OCP 3.11 compatability

twatts-redhat opened the pull request:
Have only tested playbooks/tasks that have been changed.

Commit:
ansible2.8_ocp3.11

twatts-redhat committed:
Update playbooks/ocp-rolling-os-update-and-upgrade.yml

[infra,app]_hosts --> [infra,app]-nodes

twatts-redhat committed:
Update playbooks/ocp-rolling-hosts-reboot.yml

twatts-redhat committed:
Update playbooks/ocp-rolling-services-restart.yml

twatts-redhat committed:
Update playbooks/ocp-hosts-reboot.yml

twatts-redhat committed:
Update tasks/wait-for-node-ready.yml

twatts-redhat committed:
Update playbooks/ocp-hosts-reboot.yml

itewk commented:
@twatts-redhat I am happy to merge this without trying to test myself on 2 (hopefully easy) conditions:

1) you tell me you have tested them
2) update the README with a section for each playbook (if it doesn't exist) and which version of OCP its been tested to work with

bparry02 commented on defaults/main.yml:
```diff
@@ -23,7 +23,7 @@ ocp_repositories:
     - "rhel-7-server-extras-rpms"
     - "rhel-7-server-ose-3.11-rpms"
     - "rhel-7-server-ansible-2.6-rpms"
```
Suggest removing ansible-2.6 repo from 3.11 list

twatts-redhat replied:
Removed.

bparry02 commented on playbooks/ocp-hosts-reboot.yml:
```diff
@@ -34,6 +36,7 @@
         - atomic-openshift-master-api
         - atomic-openshift-master-controllers
         - etcd
+      when: ocp_deployment_version == "3.9"
```
Has the 3.9 path been verified here? The defaults/main vars file says this variable is defined as a double/float but this looks like a string comparison. Does this succeed when expected?

twatts-redhat replied:
Have not tested the succeed when expected due to lack of available 3.9 cluster. This raises a good point, given the drift of logic between this playbook and ocp-rolling-host-update and ocp-rolling-os-update-and-upgrade

twatts-redhat replied:
Removed this when, and also the entire task which is only relevant to 3.9 services.

bparry02 commented on playbooks/ocp-rolling-hosts-reboot.yml:
```diff
@@ -12,16 +12,6 @@
   gather_facts: False
   serial: 1
   tasks:
-  - name: OCP | Rolling Reboot Hosts | Masters | Determine services
-    set_fact:
-      openshift_services:
-      - atomic-openshift-node
-      - docker
-      - atomic-openshift-master-api
-      - atomic-openshift-master-controllers
-      - etcd
-    when: openshift_services is not defined
-
```
Should this block be kept and `      when: ocp_deployment_version == "3.9"` be added to match `ocp-hosts-reboot.yml`?

twatts-redhat replied:
Removed when: ocp_deployment_version == "3.9" from ocp-hosts-reboot.yml

bparry02 commented on playbooks/ocp-rolling-hosts-reboot.yml:
```diff
@@ -36,7 +26,7 @@
       - docker
     when: openshift_services is not defined
-
+    
```
```suggestion

```

twatts-redhat replied:
Removed white space.

bparry02 commented on playbooks/ocp-rolling-os-update-and-upgrade.yml:
```diff
@@ -18,16 +18,6 @@
   gather_facts: False
   serial: 1
   tasks:
-  - name: OCP | OS Update and Upgrade | Masters | Determine services
-    set_fact:
-      openshift_services:
-      - atomic-openshift-node
-      - docker
-      - atomic-openshift-master-api
-      - atomic-openshift-master-controllers
-      - etcd
-    when: openshift_services is not defined
-
```
Should this block be kept and `      when: ocp_deployment_version == "3.9"` be added to match `ocp-hosts-reboot.yml`?

twatts-redhat replied:
Removed when: ocp_deployment_version == "3.9" from ocp-hosts-reboot.yml

bparry02 commented on playbooks/ocp-rolling-services-restart.yml:
```diff
@@ -12,16 +12,6 @@
   gather_facts: False
   serial: 1
   tasks:
-  - name: OCP | Rolling Restart Services | Masters | Determine services
-    set_fact:
-      openshift_services:
-      - atomic-openshift-node
-      - docker
-      - atomic-openshift-master-api
-      - atomic-openshift-master-controllers
-      - etcd
-    when: openshift_services is not defined
-
```
Should this block be kept and `      when: ocp_deployment_version == "3.9"` be added to match `ocp-hosts-reboot.yml`?

bparry02 commented on playbooks/ocp-rolling-services-restart.yml:
```diff
@@ -12,16 +12,6 @@
-    when: openshift_services is not defined
-
   - name: OCP | Rolling Restart Services | Masters | Include Tasks
     include_tasks: tasks/ocp-services-reboot.yml
```
I think a master-specific `ocp-services-reboot` should be created. `tasks/restart_master_services.yml` should not be part of the logic for the other nodes.

twatts-redhat replied:
Separated and moved to different tasks called by the playbook

bparry02 commented on tasks/ocp-services-reboot.yml:
```diff
@@ -12,12 +12,18 @@
+
+- name: OCP | Update OS | Restart Master API and Controllers
+  include_tasks: tasks/restart_master_services.yml
+  when: inventory_hostname in groups['masters']
```
I think a master-specific `ocp-services-reboot` should be created. `tasks/restart_master_services.yml` should not be part of the logic for the other nodes as it is currently

twatts-redhat replied:
Separated and moved to different tasks called by the playbook

bparry02 commented on tasks/os-update-and-upgrade.yml:
```diff
@@ -10,45 +10,10 @@
-      - "openshift_oc_tools_host is defined"
+- name: OCP | Update OS | Include Tasks for Masters
+  include_tasks: tasks/masters-os-update-and-upgrade.yml
+  when: inventory_hostname in groups['masters']
```
Add `- packages_need_update_result is changed` when logic to this task, and remove the block from the included task file.

twatts-redhat replied:
Made changes.

bparry02 commented on tasks/os-update-and-upgrade.yml:
```diff
@@ -10,45 +10,10 @@
-  when: packages_need_update_result is changed
+- name: OCP | Update OS | Include Tasks for Nodes
+  include_tasks: tasks/nodes-os-update-and-upgrade.yml
+  when: inventory_hostname not in groups['masters']
```
Add `- packages_need_update_result is changed` when logic to this task, and remove the block from the included task file.

twatts-redhat replied:
Made changes.

bparry02 commented on tasks/masters-os-update-and-upgrade.yml:
```diff
@@ -0,0 +1,37 @@
+    include_tasks: tasks/reboot.yml
+    vars:
+      reboot_only_if_needed: True
+
```
Add a comment here, saying that we are restarting the master services here only if the host was *not* rebooted

twatts-redhat replied:
Added comment.

bparry02 commented on tasks/wait-for-node-ready.yml:
```diff
@@ -1,18 +1,18 @@
       port: "{{ openshift_master_api_port }}"
       timeout: 600
-  when: inventory_hostname in groups['masters']
+    when: inventory_hostname in groups['masters']
```
This when statement has been indented. It now only applies to the `wait_for` instead of the the entire block. I think this change should be reverted.

```suggestion
  when: inventory_hostname in groups['masters']
```

twatts-redhat replied:
Reverted change. I believe I was thinking here that the block was on the entire task, which its not.

bparry02 commented on tasks/wait-for-node-ready.yml:
```diff
@@ -1,18 +1,18 @@
 # for master nodes wait for the master API port to be ready
 - block:
-  - name: OCP | Wait for Node Ready | Determine master API port
+  - name: OCP | Wait for Master Ready | Determine master API port
```
```suggestion
  - name: OCP | Wait for Node Ready | Determine master API port
```
I think the task naming convention that is used here is derived from the task filename, so we should leave it saying Node even though it only applies to masters.

twatts-redhat replied:
Reverted change.

bparry02 commented on tasks/wait-for-node-ready.yml:
```diff
@@ -1,18 +1,18 @@
     when: openshift_master_api_port is not defined
 
-  - name: OCP | Wait for Node Ready | Wait for master API to come back online
+  - name: OCP | Wait for Master Ready | Wait for master API to come back online
```
```suggestion
  - name: OCP | Wait for Node Ready | Wait for master API to come back online
```
I think the task naming convention that is used here is derived from the task filename, so we should leave it saying Node even though it only applies to masters.

twatts-redhat replied:
Reverted change.

bparry02 commented on tasks/wait-for-node-ready.yml:
```diff
@@ -23,17 +23,18 @@
   # Give the node three minutes to come back online.
   retries: 36
   delay: 5
+  when: inventory_hostname not in groups['masters']
```
Why was this condition added? Masters should be in a ready state too.

twatts-redhat replied:
Good catch. This should not have been added. Removed.

bparry02 commented on tasks/masters-os-update-and-upgrade.yml:
```diff
@@ -0,0 +1,37 @@
+---
```
I'd like to rename this file to show that it is only called from os-update-and-upgrade. Maybe by creating a folder. I suggest `tasks/os-update-and-upgrade/masters.yml`

twatts-redhat replied:
Made change.

bparry02 commented on tasks/nodes-os-update-and-upgrade.yml:
```diff
@@ -0,0 +1,44 @@
+---
```
I'd like to rename this file to show that it is only called from os-update-and-upgrade. Maybe by creating a folder. I suggest `tasks/os-update-and-upgrade/nodes.yml`

twatts-redhat replied:
Made change.

bparry02 reviewed (changes requested):
I see that `ocp-swap-docker-storage-driver.yml` playbook also calls `wait-for-node-ready.yml`. Was that playbook tested?

bparry02 commented:
I think it's possible we may want to remove 3.9 compatibility from these playbooks. If an older version of OCP is needed, the playbooks can be retrieved from an older release.

itewk commented:
@bparry02 i am find with just focusing on 3.11 at this point.

twatts-redhat committed:
Update defaults/main.yml

twatts-redhat committed:
Update playbooks/ocp-rolling-hosts-reboot.yml

Remove whitespace

twatts-redhat committed:
Update tasks/wait-for-node-ready.yml

twatts-redhat committed:
Rename this file to show that it is only called from os-update-and-upgrade.

twatts-redhat committed:
Update playbooks/ocp-hosts-reboot.yml

Remove compatability for 3.9 by removing task and when statement

twatts-redhat commented:
> I see that `ocp-swap-docker-storage-driver.yml` playbook also calls `wait-for-node-ready.yml`. Was that playbook tested?

This playbook was not tested. Reviewing the playbook, I don't see a reason that the logic in wait-for-node-ready should change.

twatts-redhat committed:
Separate master from node service reboot tasks

A master-specific ocp-services-reboot should be created. tasks/restart_master_services.yml should not be part of the logic for the other nodes.

twatts-redhat committed:
Update tasks/os-update-and-upgrade/masters.yml

twatts-redhat committed:
Updae os-update-and-upgrade.yml

Add - packages_need_update_result is changed when logic to this task, and remove the block from the included task file.

twatts-redhat committed:
Update tasks/wait-for-node-ready.yml

twatts-redhat commented:
Made all suggested changes. Totally removed 3.9 compatibility going forward for the given playbooks tested. Will outline those (playbooks) in a list in the README once changes are confirmed good.

itewk commented:
@twatts-redhat have you tested each of the updated playbooks?

twatts-redhat committed:
Update tasks/wait-for-node-ready.yml

twatts-redhat committed:
Move tasks/wait-for-node-ready before rescheduling the node

twatts-redhat committed:
Move tasks/wait-for-node-ready before rescheduling the node

bparry02 commented on tasks/ocp-services-reboot/masters.yml:
```diff
@@ -0,0 +1,18 @@
+
+- name: OCP | Update OS | Restart Master API and Controllers
+  include_tasks: tasks/restart_master_services.yml
+  when: inventory_hostname in groups['masters']
```
I don't think we need this condition anymore since we are in the reboot/masters.yml and it is only called for masters.

twatts-redhat replied:
Made change.

twatts-redhat commented:
> @twatts-redhat have you tested each of the updated playbooks?

I have just re-tested the playbooks again for 3.11 after making all of Bryan's suggested changes.

To enumerate, those are:
- playbooks/ocp-rolling-os-update-and-upgrade.yml 
- playbooks/ocp-rolling-hosts-reboot.yml
- playbooks/ocp-rolling-services-restart.yml
- playbooks/ocp-hosts-reboot.yml

I was unable to test the gluster role, as I don't have a cluster with that storage implemented.

Next step - updated the README

bparry02 reviewed (changes requested):
One minor change requested. The rest looks good pending README update.

twatts-redhat committed:
Update tasks/ocp-services-reboot/masters.yml

Don't need this condition anymore since we are in the reboot/masters.yml and it is only called for masters.

twatts-redhat committed:
Update README.md

twatts-redhat commented:
Updated README to reflect tested changes and OCP compatibility. There are several other playbooks that exist in this repository that are not detailed in the README as is. I also did not test every playbook in this repository for 3.11 compatibility, only the playbooks modified in this PR.

itewk reviewed (approved):
pencil eyeball review based on others comments.
